Source/WebCore/platform/text/WritingMode.h

```diff
@@ -31,233 +31,397 @@
 
 #pragma once
 
-#include "TextDirection.h"
-#include <array>
+#include <wtf/text/TextStream.h>
 
 namespace WebCore {
 
-enum class TypographicMode : bool {
-    Horizontal,
-    Vertical,
-};
+/**
+ * WritingModeData efficiently stores a writing mode and can rapidly compute
+ * interesting things about it for use in layout. It fits in 1 byte,
+ * and can be cheaply copied or passed around by value.
+ *
+ * Writing modes are computed from the CSS 'writing-mode', 'direction', and
+ * 'text-orientation' properties. See CSS Writing Modes for more information.
+ *   http://www.w3.org/TR/css-writing-modes/
+ *
+ * Be careful with the distinctions among these:
+ *   - isBidiLTR/RTL (for typesetting, i.e. within a line box)
+ *   - isInlineLeftToRight/TopToBottom (absolute physical directions)
+ *   - isLogicalLeftInlineStart (flow-relative cmp coordinate-relative)
+ *   - isLogicalLeftLineLeft (line-relative cmp coordinate-relative)
+ *   - computedTextDirection (style computation / element directionality)
+ */
 
-enum class FlowDirection : uint8_t {
-    TopToBottom,
-    BottomToTop,
-    LeftToRight,
-    RightToLeft,
+enum class WritingModeStyle : uint8_t;
+enum class TextDirection : bool;
+enum class TextOrientation : uint8_t;
+enum class FlowDirection : uint8_t;
+
+class WritingModeData final {
```

Comment:
Yeah, I figured we'd want to name one of these WritingMode and the other something else, but since I wasn't sure which way it would go I just gave them both distinguishable names so I can easily run a replacement on it. :)